File: lib/features/hotels/data/repositories/mock_hotel_repository.dart
import 'dart:async';
import '../../domain/entities/hotel.dart';
import '../../domain/entities/province.dart';
import '../../domain/repositories/hotel_repository.dart';
import '../datasources/mock_hotel_data_source.dart';

class MockHotelRepository implements HotelRepository {
  MockHotelRepository() : _hotels = MockHotelDataSource.hotels();
  List<Hotel> _hotels;
  final _controller = StreamController<List<Hotel>>.broadcast();
  @override Stream<List<Province>> watchProvinces() => Stream.value(MockHotelDataSource.provinces);
  @override Stream<List<Hotel>> watchHotels() async* { yield _hotels; yield* _controller.stream; }
  @override Future<void> setFavorite(String hotelId, bool isFavorite) async { _hotels = _hotels.map((hotel) => hotel.id == hotelId ? hotel.copyWith(isFavorite:isFavorite) : hotel).toList(growable:false); _controller.add(_hotels); }
}

File: android-app/app/src/main/java/com/hjozaty/app/data/repository/MockHotelRepository.kt
package com.hjozaty.app.data.repository

import com.hjozaty.app.core.error.AppResult
import com.hjozaty.app.core.error.UnknownError
import com.hjozaty.app.data.mock.MockHotelData
import com.hjozaty.app.domain.model.Hotel
import com.hjozaty.app.domain.model.Province
import com.hjozaty.app.domain.repository.HotelRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

/** In-memory single source of truth for rich fictional accommodation data. */
@Singleton
class MockHotelRepository @Inject constructor() : HotelRepository {
    private val hotels = MutableStateFlow(MockHotelData.hotels)
    override fun observeProvinces(): Flow<List<Province>> = kotlinx.coroutines.flow.flowOf(MockHotelData.provinces)
    override fun observeHotels(): Flow<List<Hotel>> = hotels
    override fun observeFeaturedHotels(): Flow<List<Hotel>> = hotels.map { list -> list.filter(Hotel::isFeatured) }
    override fun observeRecommendedHotels(): Flow<List<Hotel>> = hotels.map { list -> list.filter(Hotel::isRecommended) }
    override fun observeHotelsByProvince(provinceId: String): Flow<List<Hotel>> = hotels.map { list -> list.filter { it.provinceId == provinceId } }
    override fun observeHotelById(hotelId: String): Flow<Hotel?> = hotels.map { list -> list.firstOrNull { it.id == hotelId } }
    override suspend fun setFavorite(hotelId: String, isFavorite: Boolean): AppResult<Unit> {
        val current = hotels.value
        if (current.none { it.id == hotelId }) return AppResult.Failure(UnknownError)
        hotels.value = current.map { if (it.id == hotelId) it.copy(isFavorite = isFavorite) else it }
        return AppResult.Success(Unit)
    }
}
